### PR TITLE
Centralize CPU features, fix `iree_cpu_lookup_data_by_key`, add a `iree-cpuinfo` tool

### PR DESCRIPTION
**Pull request by bjacob**
The starting point of this PR was that I realized that I had forgotten to update `iree_cpu_lookup_data_by_key`.

That sent me looking for a way to have it automatically sync'd with the rest of the code, resulting in the present `.inl` approach.

As a bonus, I added a `iree-cpuinfo` tool, also automatically sync'd with the `.inl` and just a few lines of impl. This is a quick debugging tool allowing to answer quickly the question "which CPU features are we actually detecting on this machine". Obviously the point is not provide the best CPU feature detection tool, the point is that it's by construction bug-for-bug equivalent to IREE's actual CPU feature detection.

There are some more unification opportunities beyond this PR. Ukernel tests have a `iree_uk_test_cpu_features_str` that could use this `.inl` and drop its custom impl, or even better, ukernel tests could be restructured to go from string key to enum bit, instead of currently going in the opposite direction.

I have had misgivings about `iree_cpu_lookup_data_by_key` living in core when it's not yet clear that it will ever be used outside of tests/tools. But thanks to the `IREE_ARCH_ENUM` introduced here, the arch condition gets evaluated at compile time so each arch pays only for what it uses, and even x86-64 has `iree_cpu_lookup_data_by_key` size 1178 bytes, while arm_64 has `iree_cpu_lookup_data_by_key` size 204 bytes.

### CODE REVIEW

**Comment by bjacob** [comment]
Output of `iree-cpuinfo` on my desktop:

```
sse3                 1
ssse3                1
sse4.1               1
sse4.2               1
sse4a                0
avx                  1
fma                  1
fma4                 0
xop                  0
f16c                 1
avx2                 1
avx512f              1
avx512cd             1
avx512vl             1
avx512dq             1
avx512bw             1
avx512ifma           0
avx512vbmi           0
avx512vpopcntdq      0
avx512vnni           0
avx512vbmi2          0
avx512bitalg         0
avx512bf16           0
avx512fp16           0
amx-tile             0
amx-int8             0
amx-bf16             0
```

On my Android test device:

```
dotprod              1
i8mm                 1
```

**Comment by bjacob** [comment]
By the way, to answer the question "how much code size do we save by having our own minimalistic cpu-querying code".

All the detection code is in `iree_cpu_initialize`, and that's the only significant symbol besides that `iree_cpu_lookup_data_by_key` discussed above (which, again, could move to tools if needed).

On x86-64, `iree_cpu_initialize` is 704 bytes. Not too bad to perform detection of 27 features. [Disassembly](https://gist.github.com/bjacob/97c82b8582d3810390e591cbbacf2922).

On arm64, `iree_cpu_initialize` is 80 bytes. That's 20 instructions. That's admittedly because we are currently only checking 2 features (`dotprod` and `i8mm`) but also because `getauxval` is so easy to use. Here's the full arm64 disassembly:

```
0000000000023038 <iree_cpu_initialize>:
   23038: fd 7b be a9   stp     x29, x30, [sp, #-32]!
   2303c: f4 4f 01 a9   stp     x20, x19, [sp, #16]
   23040: fd 03 00 91   mov     x29, sp
   23044: 00 e4 00 6f   movi    v0.2d, #0000000000000000
   23048: 94 02 00 f0   adrp    x20, 0x76000 <iree_cpu_lookup_data_by_key+0xb4>
   2304c: 94 02 28 91   add     x20, x20, #2560
   23050: 00 02 80 52   mov     w0, #16
   23054: 80 02 01 ad   stp     q0, q0, [x20, #32]
   23058: 80 02 00 ad   stp     q0, q0, [x20]
   2305c: 95 35 01 94   bl      0x706b0 <getauxval@plt>
   23060: f3 03 00 aa   mov     x19, x0
   23064: 40 03 80 52   mov     w0, #26
   23068: 92 35 01 94   bl      0x706b0 <getauxval@plt>
   2306c: 08 fc 4c d3   lsr     x8, x0, #12
   23070: 08 01 7f 92   and     x8, x8, #0x2
   23074: 68 52 54 b3   bfxil   x8, x19, #20, #1
   23078: 88 02 00 f9   str     x8, [x20]
   2307c: f4 4f 41 a9   ldp     x20, x19, [sp, #16]
   23080: fd 7b c2 a8   ldp     x29, x30, [sp], #32
   23084: c0 03 5f d6   ret
```

**Comment by benvanik** [comment]
(will review in a bit, but iree_cpu_lookup_data_by_key is not tool only - that's how the compiler can query for features at runtime on the host - of course, if you think you'll never want that or only want certain things we can hide it, but that'd mean you wouldn't be able to change workgroup counts/etc based on cpu features/cache sizes/etc)

**Comment by bjacob** [comment]
I see the idea but we seem to have already taken a different, equivalent path in `host` compile:

* The `host` CPU feature detection is done here: https://github.com/openxla/iree/blob/7c3e6a5047b663b2f44400dc5622396e11021b22/compiler/src/iree/compiler/Dialect/HAL/Target/LLVM/LLVMTargetOptions.cpp#L35 
* The resulting LLVM target attribute strings are used in this kind of logic: https://github.com/openxla/iree/blob/7c3e6a5/compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUMaterializeEncodingPass.cpp#L31-L85

**Comment by bjacob** [comment]
(Just to clarify, all this is a tangent here, not directly relevant to the PR at hand. We're at most discussing a possible follow-up).

**Comment by benvanik** [comment]
What the compiler is doing is different - this is about the host-side runtime queries we can emit for changing allocations/selecting alternate data formats/etc. The only way things work today is because we've been riding with unportable code all this time and hardcoding things in the compiler we shouldn't be :)

The more things we can keep in the dispatches the better (exact tile size/etc) but for anything that will change program flow or workgroup counts we'll need to be able to emit runtime queries. Ideally as few things as possible need that though - workgroup count should ideally be based mostly on problem size and processor count and not how each workgroup processes its individual local slice of the problem. If we never end up needing architectural info (cache sizes, vector widths, supported instruction sets, etc) on the host that'd be the best outcome, but I think someone really needs to drive the work to un-hardcode the compiler side assuming that if a flag is passed to iree-compile that means anything about the runtime environment so we can see if that holds. Fingers crossed it does :)

**Review by benvanik — APPROVED** [review]

**Comment by bjacob** [comment]
I see, thanks for the explanation!

**Comment by dcaballe** [comment]
We are also checking some CPU features during code generation: https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/Utils/Utils.h#L77-L93. At that time, we decided that we would have our own way to query/represent them in C/C++ (vs using LLVM's TTI) as long as we only needed a handful number of them. Looking at this PR, it looks like this is no longer the case so I think we should revisit that decision. At the very least, we should have a single API that works for codegen and ukernels. Have you considered using TTI for this instead of having our own representation?

**Comment by benvanik** [comment]
As mentioned this is for the runtime and not for the compiler so TTI isn't available - but even in the compiler where it is we shouldn't be relying on fixed information unless absolutely required. Emitting a switch at runtime that choses one 200 byte block of instructions vs another in the one place that needs it is the only way we end up with reasonable binary sizes when trying to support more than just one particular architecture. We already have the provisions for doing this both on the host (workgroup size/count with `hal.device.query`) and within the device dispatches (`DispatchABI::loadProcessorID` and `DispatchABI::loadProcessorData`) - those all return SSA values for the runtime parameters backed by the information in this PR and can be used to switch code paths/tune/etc.

Basically, if our compiler is built to assume fixed architectural information at the top level and the user wants to target more than one architecture the only solution is to compile everything multiple times and bundle up multiple copies. That's not good :) We need to move away from that by parameterizing as much as we can in the compiler such that where we make decisions is either at the leafs in the compiler (call block function X or block function Y based on arch) or runtime queries (compile both block function X and Y and then switch at runtime).

**Comment by dcaballe** [comment]
Thanks for the clarification, Ben! I'm not questioning the approach but more asking for a unified way to represent target features in IREE. We should have a single way, regardless of whether the queries happen at runtime or compile time. If that means having our own representation, that's fine, as long as we can also use it for all the use cases we have today. 

TTI as a whole can be initialized to arbitrary targets and it's used for AOT, JIT, cross-compilation, etc. so I understand that the main concern is creating a dependency between the runtime and the LLVM backend? TTI also uses different components, such as TargetMachine, which is the one that would fit here (e.g., x86 implementation https://llvm.org/doxygen/classllvm_1_1X86Subtarget.html). However, if the answer is that we should implement our own, that's fine as well, but we have to make sure the solution also works for the current codegen cases in [main/compiler/src/iree/compiler/Codegen/Utils/Utils.h#L77-L93](https://github.com/openxla/iree/blob/main/compiler/src/iree/compiler/Codegen/Utils/Utils.h?rgh-link-date=2023-03-07T19%3A57%3A17Z#L77-L93). We should avoid having two parallel ways to represent target features in IREE.
